### ISSUE TITLE
Fix the non-converging issue of SVD on GPU for large matrices

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#64237]
## 🚀 Feature

SVD on GPU for a large matrix (usually with size > 1024) or an ill-conditioned matrix may throw runtime error because of not converging well. 

SVD on GPU currently uses the iterative Jacobi method of `cusolverDn<T>gesvdj` (note the trailing `j`). The pytorch implementation takes cusolver default values of `max sweeps = 100` and `tolerance = machine accuracy`. The cusolver `gesvdj` engine stops when either the accuracy or max sweeps is achieved. From the cusolver [doc](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-gesvdj), usually ~15 sweeps would be enough.

The `gesvdj` method works well and is much faster than the QR-based method of `cusolverDn<T>gesvd` for small size matrix (benchmark https://github.com/pytorch/pytorch/pull/48436#issuecomment-733430414). However, for large size matrix, we've received several reports with runtime errors because SVD is not converged well, e.g. https://github.com/pytorch/pytorch/issues/28293#issuecomment-850822958

To fix this issue, there are several approaches

1. Expose the `max sweeps` and `tolerance` of cusolver `gesvdj` to user, so that people can set them to stricter values as needed.
_Problem_: what should the user interface look like? Extra kwargs for `torch.linalg.svd`? Note that the CPU LAPACK method doesn't have this issue and won't use those parameters.

2. Add `gesvd` method to pytorch. 
_Problem_: what should the user interface look like? The entrance of SVD is only `torch.linalg.svd` and how should user change the backend engine? Extra kwargs or extra python functions?

3. Add `gesvd` method to pytorch, and use `gesvd` as a fallback when `gesvdj` doesn't converge. 
_Problem_: this would make SVD performance much worse in those cases.

Note that the QR-based `gesvd` method always converges. This issue doesn't affect the batched `gesvdjBatched` method which only takes matrices with size <= 32 and converges well in nearly all cases.

**Comment by Lezcano** [comment]
In my opinion, the third option is the best one.

I used to face this problem very rarely (say while training a network once every 10 epochs) and it would make the training fail, which was very annoying. As such, if we just launch `gesvd` as a fallback, the amortised cost will be close to zero, and all this would be completely transparent to the user.

Note that the solution I used to do (and one they propose in the mentioned post) is to just have a `try-catch`, and in the catch, one perturbs the original matrix with some noise and computes the SVD of that one. This effectively also incurs in computing two SVDs, but, again, this error is so infrequent that it's fine.

**Comment by Lezcano** [comment]
There is a similar problem with `eigh` https://github.com/pytorch/pytorch/issues/59720 (and I suspect that also with `eig`), so we should see if we have similar fallback algorithms for these other two. I'd assume that there is one for `eigh`, but `eig` is infamous for not being the most stable of functions...

**Comment by Lezcano** [comment]
This is a similar one, but it fails silently: https://github.com/pytorch/pytorch/issues/24466
For this second one we should probably just mention in the docs that working with floating precision may not be enough for these algorithms when the matrices are ill-conditioned.

**Comment by mruberry** [comment]
Per offline discussion:

- the fallback sounds great 
- a separate issue to add a "driver"-like argument (like lstsq has) would be interesting (we should consider giving them meaningful names and their algorithmic names, too)
  - alternatively the keyword might be something like "exact"
- we don't think we should add the more precise CUDA-specific kwargs to linalg.svd at this time (but maybe a "lower level" function in the future could expose them?)  
  - structure work may be able to expose these options in the future, too
- the fallback to the alternative "more precise" algorithm must be carefully documented because users might complain about us being different from other libraries and confused about the performance of linalg.svd

**Comment by o-alexandre-felipe** [comment]
Do you think a parallel QR iteration for tridiagonal systems would help any of this?
As far as I know the method used for large matrices is the divide and conquer, in which sub matrices are diagonalized and then, combined with a rank-1 perturbation.
I have read that the QR iteration method is still the best in a serial computer.
I have framed the sub diagonal elimination step of the QR iteration as an associative operation (in an augmented space).
If the serial elimination of the sub diagonal has complexity c1 * N, in a serial computer, this will have a complexity c2 * log(N) in a fully parallel computer, or c2 * N * log(N) / P, when we have P processors.
